Minor libpacemaker and libpe_status improvements

nrwahl2 opened the pull request:
If I don't get any farther on [T672](https://projects.clusterlabs.org/T672) tonight, at least we have these.

I propose we drop support for colocations. Just let resources run wherever they want relative to other resources, who cares anymore.

nrwahl2 commented on lib/pacemaker/pcmk_sched_resource.c:
```diff
@@ -384,13 +384,11 @@ add_assigned_resource(pe_node_t *node, pe_resource_t *rsc)
-bool
-pcmk__finalize_assignment(pe_resource_t *rsc, pe_node_t *chosen, bool force)
+static bool
+assign_resource(pe_resource_t *rsc, pe_node_t *chosen, bool force)
```
I took "finalize" out of the name proactively, because we'll need it to be fully reversible in order to fix [T489](https://projects.clusterlabs.org/T489).

There's currently no way to undo an assignment if `chosen == NULL` or if we fail to find an available node. `pcmk__finalize_assignment()` sets `rsc->next_role` to Stopped in that case and updates its existing actions. `pcmk__unassign_resource()` has no way to undo that.

I have another commit queued up locally that adds a flag for this. I'm saving it until the T489 fix is ready, and that in turn depends on T672.

nrwahl2 commented on lib/pacemaker/pcmk_sched_resource.c:
```diff
@@ -488,14 +484,19 @@ pcmk__assign_resource(pe_resource_t *rsc, pe_node_t *node, bool force)
+
         if (rsc->allocated_to != NULL) {
-            changed = true;
+            old = pe__copy_node(rsc->allocated_to);
```
We could just save the old node details and compare to the new one. I'd prefer to use `pe__copy_node()` so that we can use `pe__same_node()` for clarity.

kgaillot commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -172,19 +172,22 @@ cmp_primary_priority(gconstpointer a, gconstpointer b)
               colocation->id, colocation->dependent->id,
               colocation->primary->id, colocation->node_attribute,
-              colocation->score);
+              colocation->score, rsc->id);
```
pe_rsc_trace()

nrwahl2 replied:
Thanks, hitting it here and in the other three places

kgaillot commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -195,23 +198,32 @@ pcmk__add_this_with(GList **list, const pcmk__colocation_t *colocation)
+    CRM_ASSERT((list != NULL) && (rsc != NULL));
+
+    pcmk__if_tracing(
+        {}, // Add each colocation individually if tracing
```
"Always add" is a little clearer

nrwahl2 replied:
Agreed

kgaillot reviewed (commented):
looking good

nrwahl2 commented:
Updated to address review:
* "static-ize" commit becomes "de-functionize" commit, and update `cts-cli` tests to reflect new function name
* Use `pe_rsc_trace()` and improve comments in logging commit

No other changes

kgaillot commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -195,23 +199,34 @@ pcmk__add_this_with(GList **list, const pcmk__colocation_t *colocation)
+                pe_rsc_trace(rsc,
+                             "Copying %u 'this with' colocations to new list "
+                             "for %s",
+                             g_list_length(addition), rsc->id);
```
We know we're not tracing here ... but someone can use PCMK_trace_formats so this is still theoretically useful

nrwahl2 replied:
I'll update the comment real quick. I meant it in contrast to the "if tracing" above -- so in the sense of "not tracing, so trivial case for efficiency"

nrwahl2 replied:
OH, the `pe_rsc_trace()`... crap

nrwahl2 replied:
We could either drop it (preferable) or make it debug (seems inappropriate). It seems silly to keep it on the off-chance someone is using a matching `PCMK_trace_formats`, honestly.

kgaillot replied:
Just drop it

nrwahl2 commented:
small update:
before:
```
    if (rsc->allocated_to != NULL) {
        changed = !pe__same_node(rsc->allocated_to, node);
    }
```

after:
```
    if (rsc->allocated_to != NULL) {
        changed = !pe__same_node(rsc->allocated_to, node);
    } else {
        changed = (node != NULL);
    }
```

kgaillot commented:
> ```
>     if (rsc->allocated_to != NULL) {
>         changed = !pe__same_node(rsc->allocated_to, node);
>     } else {
>         changed = (node != NULL);
>     }
> ```

`pe__same_node()` handles NULL, so you can just drop the if-else

kgaillot commented:
> > ```
> >     if (rsc->allocated_to != NULL) {
> >         changed = !pe__same_node(rsc->allocated_to, node);
> >     } else {
> >         changed = (node != NULL);
> >     }
> > ```
> 
> `pe__same_node()` handles NULL, so you can just drop the if-else

Not necessary for this PR though

nrwahl2 commented:
> > ```
> >     if (rsc->allocated_to != NULL) {
> >         changed = !pe__same_node(rsc->allocated_to, node);
> >     } else {
> >         changed = (node != NULL);
> >     }
> > ```
> 
> `pe__same_node()` handles NULL, so you can just drop the if-else

You'd think so :)

In the `else` case, `rsc->allocated_to == NULL`, so `pe__same_node()` would **always** return false. The assignment is changing if `(rsc->allocated_to == NULL) && (node != NULL)` at this point.
